### PR TITLE
Making ShadowMap constructor private

### PR DESCRIPTION
**Pull request by rahwang**
See #4010

### CODE REVIEW

**Comment by lilleyse** [comment]
I'm not sure if this is possible to do, but can the constructor be private but `ShadowMap` still be visible in the documentation? Shadows have a few settings like `darkness`, `maximumDistance`, `size`, etc which would be nice to have in the documentation still.

**Comment by pjcozzi** [comment]
> I'm not sure if this is possible to do, but can the constructor be private but ShadowMap still be visible in the documentation?

`@internalConstructor` - not sure how well supported it is now - see [ModelMesh](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/ModelMesh.js) for an example.

> Shadows have a few settings like darkness, maximumDistance, size, etc which would be nice to have in the documentation still.

How do users get to the shadow map object?  Through the scene?

We could keep this class public with the private constructor and make any other members private as needed.

**Comment by lilleyse** [comment]
`viewer.shadowMap` is the standard way, but its also accessible from scene.

**Comment by rahwang** [comment]
How's this, @lilleyse ? Should normalOffset be visible since its mentioned?

**Comment by lilleyse** [comment]
Looking at jsdoc documentation `@internalConstructor` isn't supported, and I can't find another way of doing this...

Whether the constructor can be hidden or not, I think its still a good idea to keep the parameters since they are helpful for anyone coding cesium.

Maybe we just need a note that it is not recommend to construct a `ShadowMap` directly since Cesium does not support multiple light sources.

I don't see a problem with instantiating with `scene` instead of `context`, and just get the `context` with `scene.context`. Passing in scene is starting to get more common now with height references (like Model). This will have breaking changes though here and in Cesium Pro. Technically I think this can all be done without sending in context or scene, but will require bigger changes. 

> Should normalOffset be visible since its mentioned?

Hm, I suppose there should be a setting for this since it is mentioned. The normal offset is controlled by setting `_terrainBias`, `_primitiveBias`, and `_pointBias`'s `normalOffset` property.

**Comment by lilleyse** [comment]
Oh so Cesium adds `internalConstructor` as a synonym for `class` in `cesiumTags.js`.

**Comment by lilleyse** [comment]
Overall adding that warning comment and using `scene` instead of `context` are what I would do for the time being. @pjcozzi what do you think about passing in `scene'?

**Comment by pjcozzi** [comment]
> Overall adding that warning comment and using scene instead of context are what I would do for the time being. @pjcozzi what do you think about passing in scene?

Yes, more precisely
* Pass in `scene`, instead of `context`
* Mark the constructor `internalConstructor`.  Include the "warning comment" just like the example I pointed to in [ModelMesh](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/ModelMesh.js#L11) - this should point the reader to [Viewer.shadowMap](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Widgets/Viewer/Viewer.js#L986) to get the object
* @lilleyse please evaluate if any members of `ShadowMap` should now be marked `@private`

**Comment by lilleyse** [comment]
> @lilleyse please evaluate if any members of ShadowMap should now be marked @private

All looks good.

**Comment by pjcozzi** [comment]
@rahwang any update here?

**Comment by rahwang** [comment]
Changes made, @pjcozzi . Now using `scene` instead of `context`.

File: Source/Scene/ShadowMap.js [hunk 1]
```diff
@@ -107,37 +107,28 @@ define([
 
     /**
      * Creates a shadow map from the provided light camera.
+     * <p>
+     *     Use {@link Viewer#shadowMap}
```
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: It's fine to remove the indentation for this line.

Expand on the description a bit. Maybe `Use {@link Viewer#shadowMap} to get the scene's shadow map originating from the sun. In general do not construct directly.`

File: Source/Scene/ShadowMap.js [hunk 2]
```diff
@@ -107,37 +107,28 @@ define([
...
-     * @param {Camera} options.lightCamera A camera representing the light source.
-     * @param {Boolean} [options.enabled=true] Whether the shadow map is enabled.
-     * @param {Boolean} [options.isPointLight=false] Whether the light source is a point light. Point light shadows do not use cascades.
-     * @param {Boolean} [options.pointLightRadius=100.0] Radius of the point light.
-     * @param {Boolean} [options.cascadesEnabled=true] Use multiple shadow maps to cover different partitions of the view frustum.
-     * @param {Number} [options.numberOfCascades=4] The number of cascades to use for the shadow map. Supported values are one and four.
-     * @param {Number} [options.maximumDistance=5000.0] The maximum distance used for generating cascaded shadows. Lower values improve shadow quality.
-     * @param {Number} [options.size=2048] The width and height, in pixels, of each shadow map.
-     * @param {Boolean} [options.softShadows=false] Whether percentage-closer-filtering is enabled for producing softer shadows.
-     * @param {Number} [options.darkness=0.3] The shadow darkness.
```
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: It's still nice to keep these parameter descriptions even if the shadow map isn't meant to be constructed directly.

File: Source/Scene/ShadowMap.js [hunk 3]
```diff
@@ -107,37 +107,28 @@ define([
...
+     * @internalConstructor
      *
      * @exception {DeveloperError} Only one or four cascades are supported.
      *
      * @demo {@link http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Shadows.html|Cesium Sandcastle Shadows Demo}
+     *
```
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: Remove empty line.

File: Source/Scene/ShadowMap.js [hunk 4]
```diff
@@ -149,6 +140,8 @@ define([
 
         this._enabled = defaultValue(options.enabled, true);
         this._softShadows = defaultValue(options.softShadows, false);
+        this._normalOffset = defaultValue(options.normalOffset, true);
```
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: When the parameters are added back to the doc, include `normalOffset`.

File: Source/Scene/ShadowMap.js [hunk 5]
```diff
@@ -149,6 +140,8 @@ define([
...
+        this._normalOffsetScale = defaultValue(options.normalOffsetScale, 0.1);
```
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: I would not expose this right now as an option or a getter/setter. The default values used before are tuned for the different bias modes.

File: Source/Scene/ShadowMap.js [hunk 6]
```diff
@@ -385,6 +379,40 @@ define([
...
+         * @type {Boolean}
+         * @default true
+         */
+        normalOffset : {
+            get : function() {
+                return this._normalOffset;
+            },
+            set : function(value) {
+                this.dirty = this._normalOffset !== value;
+                this._normalOffset = value;
```
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: This setter should also set `this._terrainBias.normalOffset`, `this._primitiveBias.normalOffset`, and `this._pointBias.normalOffset`,

Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: Also add a test for this. It should be fine to set `normalOffset` to false and expect the `normalOffset` to be false for all the bias modes. I don't think this requires any rendering as part of the test.

**Comment by lilleyse** [comment]
Looks good!

**Pull request merged**